src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -2598,28 +2602,69 @@ impl Step for Crate {
         let target = self.target;
         let mode = self.mode;
 
-        // See [field@compile::Std::force_recompile].
-        builder.ensure(compile::Std::force_recompile(compiler, compiler.host));
-
-        if builder.config.build != target {
-            builder.ensure(compile::Std::force_recompile(compiler, target));
-            builder.ensure(RemoteCopyLibs { compiler, target });
-        }
-
         // If we're not doing a full bootstrap but we're testing a stage2
         // version of libstd, then what we're actually testing is the libstd
         // produced in stage1. Reflect that here by updating the compiler that
         // we're working with automatically.
         let compiler = builder.compiler_for(compiler.stage, compiler.host, target);
 
-        let mut cargo = builder::Cargo::new(
-            builder,
-            compiler,
-            mode,
-            SourceType::InTree,
-            target,
-            builder.kind.as_str(),
-        );
+        let mut cargo = if builder.config.cmd.miri() {
+            if builder.top_stage == 0 {
+                eprintln!("ERROR: `--miri` requires stage 1 or higher");
+                std::process::exit(1);
+            }
+            if builder.kind.as_str() != "test" {
+                eprintln!("ERROR: `--miri` only works with `test` mode");
+                std::process::exit(1);
+            }
+
+            // Prepare sysroots
+            builder.ensure(compile::Std::new(compiler, compiler.host));
+            let sysroot = builder.sysroot(compiler);
+            let miri_sysroot = Miri::build_miri_sysroot(builder, compiler, target);
+
+            // Build `cargo miri test` command
+            let mut cargo = builder::Cargo::new(
+                builder,
+                compiler,
+                mode,
+                SourceType::InTree,
+                target,
+                "miri-test",
+            );
+            cargo.env("MIRI_SYSROOT", &miri_sysroot);
+            cargo.env("MIRI_HOST_SYSROOT", &sysroot);
+            // See `Miri` tests for why
+            if builder.doc_tests != DocTests::No {
+                cargo.env("RUSTDOC", builder.rustdoc(compiler));
+            }
```

Comment:
This logic is duplicated between here and the `test::Miri` step. It should probably be done somewhere inside `Cargo::new`.